apps/web/src/app/(profile)/[username]/_components/dashboard/_actions.ts

```diff
@@ -167,71 +173,51 @@ export async function getSolvedChallenges(userId: string) {
   };
 }
 
-export interface BadgeInfo {
-  // eslint-disable-next-line @typescript-eslint/sort-type-constituents
-  slug: 'aot-2023-bronze' | 'aot-2023-silver' | 'aot-2023-gold' | 'aot-2023-platinum';
+export interface Badges<T> {
+  slug: T;
   name: string;
+  shortName: string;
 }
 
-export async function getBadges(userId: string): Promise<BadgeInfo[]> {
-  const badges: BadgeInfo[] = [];
-
-  const holidayTrack = await prisma.track.findFirst({
-    where: {
-      slug: 'advent-of-typescript-2023',
-    },
-    include: {
-      trackChallenges: {
-        orderBy: {
-          orderId: 'asc',
-        },
-        include: {
-          challenge: {
-            include: {
-              submission: {
-                where: {
-                  userId,
-                },
-              },
-            },
-          },
-        },
-      },
-      enrolledUsers: {
-        where: {
-          id: userId,
-        },
-        select: {
-          id: true,
-        },
-      },
-    },
-  });
+type BadgeTypes = AotBadges | DifficultyBadges | SolutionBadges;
+export type BadgeObj<T> = {
+  [key in BadgeTypes]?: {
+    slug: key;
+    name: string;
+    shortName: string;
+  };
+};
 
-  const numberOfAttemptedHolidayChallenges =
-    holidayTrack?.trackChallenges.filter((trackChallenge) => {
-      return (trackChallenge.challenge.submission?.length ?? 0) > 0;
-    }).length ?? 0;
+export type AllBadges = Badges<BadgeTypes>;
+export type AllBadgeObjs = BadgeObj<BadgeTypes>;
 
-  if (numberOfAttemptedHolidayChallenges > 0) {
-    badges.push({ slug: 'aot-2023-bronze', name: 'Advent of TypeScript 2023 Bronze' });
-  }
+export type BadgeFn = ({ userId, badges }: { userId: string; badges: AllBadgeObjs }) => Promise<AllBadgeObjs>;
 
-  const numberOfCompletedHolidayChallenges =
-    holidayTrack?.trackChallenges.filter((trackChallenge) => {
-      return trackChallenge.challenge.submission?.some((submission) => submission.isSuccessful);
-    }).length ?? 0;
+export async function getBadges(userId: string): Promise<AllBadgeObjs> {
+  let badges: AllBadgeObjs = {};
 
-  if (numberOfCompletedHolidayChallenges >= 5) {
-    badges.push({ slug: 'aot-2023-silver', name: 'Advent of TypeScript 2023 Silver' });
-  }
+  // calculate badges user has achieved
+  const badgeCalculations: BadgeFn[] = [
+    adventBadgesFn,
+    difficultyBadgesFn,
+    sharedSolutionsBadgesFn,
+  ];
 
-  if (numberOfCompletedHolidayChallenges >= 15) {
-    badges.push({ slug: 'aot-2023-gold', name: 'Advent of TypeScript 2023 Gold' });
+  for (const badgeFn of badgeCalculations) {
+    badges = await badgeFn({ userId, badges });
   }
 
-  if (numberOfCompletedHolidayChallenges >= 25) {
-    badges.push({ slug: 'aot-2023-platinum', name: 'Advent of TypeScript 2023 Platinum' });
+  // retrieve current awarded badges
+  const userBadges: { slug: string; }[] = await prisma.$queryRaw`SELECT badge_name AS slug FROM UserBadges WHERE user_id=${userId}`;
+
+  // award missing badges
```

Comment:
I think the awarding of badges should be done on submission.

This endpoint can be changed to just grab the existing badges.